src/misc/messages.cpp

```diff
@@ -21,39 +21,61 @@
 
 #include "dosbox.h"
 
+#include <clocale>
 #include <cstdio>
 #include <cstdlib>
 #include <cstring>
 #include <deque>
-#include <clocale>
+#include <map>
 
 #include "std_filesystem.h"
 #include <string>
 #include <unordered_map>
 
+#include "ansi_code_markup.h"
+#include "checks.h"
 #include "control.h"
 #include "cross.h"
 #include "fs_utils.h"
-#include "string_utils.h"
 #include "setup.h"
+#include "string_utils.h"
 #include "support.h"
-#include "ansi_code_markup.h"
 
 #define LINE_IN_MAXLEN 2048
 
+CHECK_NARROWING();
+
 class Message {
 private:
 	std::string markup_msg = {};
 	std::string rendered_msg = {};
 
+	std::map<uint16_t, std::string> markup_cp_msg   = {};
+	std::map<uint16_t, std::string> rendered_cp_msg = {};
+
+	bool is_utf8_msg = false;
+
 public:
 	Message() = delete;
-	Message(const char *markup) { Set(markup); }
+	Message(const char *markup, const bool is_utf8)
+	{
+		Set(markup, is_utf8);
+	}
 
 	const char *GetMarkup()
 	{
 		assert(markup_msg.length());
-		return markup_msg.c_str();
+
+		if (!is_utf8_msg)
+			return markup_msg.c_str();
+
+		const uint16_t cp = UTF8_GetCodePage();
+		if (markup_cp_msg[cp].empty()) {
+			if (!UTF8_RenderForDos(markup_msg, markup_cp_msg[cp], cp))
+			LOG_WARNING("LANG: Problem rendering markup string");
+			assert(markup_cp_msg[cp].length());
+		}
+		return markup_cp_msg[cp].c_str();
```

Comment:
So for example I find the use of the name `cp` acceptable, and even preferable. It's a short scope, just a few lines, and it's clear from `UTF8_GetCodePage` that `cp` stores the code of a code page 👍🏻

Just to further explain my previous comment (short var names ok for small scopes, but the bigger the scope, the more important it is to have fully descriptive names).

Reply:
Also note the formatting here, which is a minor issue.

Reply:
Changed naming, applied a format committing script.